File: experimental-plugins/agentic-workflows/src/agentic_workflows/parser.py
"""YAML workflow parser."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

import yaml


class StepType(Enum):
    """Types of workflow steps."""

    PROMPT = "prompt"
    COMMAND = "command"
    PARALLEL = "parallel"
    CONDITIONAL = "conditional"
    RECURRING = "recurring"
    WAIT_FOR_HUMAN = "wait-for-human"


@dataclass
class Variable:
    """Workflow variable definition."""

    name: str
    type: str
    required: bool = True
    default: Any = None
    description: str = ""


@dataclass
class GitSettings:
    """Git configuration for a workflow."""

    enabled: bool = False
    worktree: bool = False
    auto_commit: bool = True
    auto_pr: bool = True
    branch_prefix: str = "agentic"


@dataclass
class WorkflowSettings:
    """Workflow-level settings."""

    max_retry: int = 3
    timeout_minutes: int = 60
    track_progress: bool = True
    autofix: str = "none"
    terminal_output: str = "base"
    bypass_permissions: bool = False
    git: GitSettings = field(default_factory=GitSettings)


@dataclass
class StepDefinition:
    """Definition of a workflow step."""

    name: str
    type: StepType
    prompt: str | None = None
    agent: str | None = None
    command: str | None = None
    args: dict[str, Any] = field(default_factory=dict)
    steps: list["StepDefinition"] = field(default_factory=list)
    merge_strategy: str = "wait-all"
    merge_mode: str = "independent"
    condition: str | None = None
    then_steps: list["StepDefinition"] = field(default_factory=list)
    else_steps: list["StepDefinition"] = field(default_factory=list)
    max_iterations: int = 3
    until: str | None = None
    message: str | None = None
    polling_interval: int = 15
    on_timeout: str = "abort"
    model: str = "sonnet"
    step_timeout_minutes: int | None = None
    step_max_retry: int | None = None
    on_error: str = "retry"
    checkpoint: bool = False
    depends_on: str | None = None


@dataclass
class OutputDefinition:
    """Definition of a workflow output."""

    name: str
    template: str
    path: str
    when: str = "completed"


@dataclass
class WorkflowDefinition:
    """Complete workflow definition."""

    name: str
    version: str
    description: str
    settings: WorkflowSettings
    variables: list[Variable]
    steps: list[StepDefinition]
    outputs: list[OutputDefinition]


class WorkflowParseError(Exception):
    """Error parsing workflow YAML."""

    pass


class WorkflowParser:
    """Parser for YAML workflow definitions."""

    def __init__(self, base_path: Path | None = None):
        self.base_path = base_path or Path.cwd()

    def parse_file(self, path: Path) -> WorkflowDefinition:
        """Parse workflow from YAML file."""
        if not path.exists():
            raise WorkflowParseError(f"Workflow file not found: {path}")

        try:
            with open(path, encoding="utf-8") as f:
                data = yaml.safe_load(f)
        except yaml.YAMLError as e:
            raise WorkflowParseError(f"Invalid YAML: {e}") from e

        self.base_path = path.parent
        return self._parse_dict(data)

    def parse_string(self, content: str) -> WorkflowDefinition:
        """Parse workflow from YAML string."""
        try:
            data = yaml.safe_load(content)
        except yaml.YAMLError as e:
            raise WorkflowParseError(f"Invalid YAML: {e}") from e
        return self._parse_dict(data)

    def _parse_dict(self, data: Any) -> WorkflowDefinition:
        """Parse workflow from dictionary."""
        if not isinstance(data, dict):
            raise WorkflowParseError("Workflow must be a dictionary")

        name = self._required(data, "name")
        version = data.get("version", "1.0")
        description = data.get("description", "")

        settings = self._parse_settings(data.get("settings", {}))
        variables = [self._parse_variable(v) for v in data.get("variables", [])]
        steps = [self._parse_step(s) for s in data.get("steps", [])]
        outputs = [self._parse_output(o) for o in data.get("outputs", [])]

        return WorkflowDefinition(
            name=name,
            version=version,
            description=description,
            settings=settings,
            variables=variables,
            steps=steps,
            outputs=outputs,
        )

    def _required(self, data: dict[str, Any], key: str) -> Any:
        """Get required field or raise error."""
        if key not in data:
            raise WorkflowParseError(f"Missing required field: {key}")
        return data[key]

    def _parse_settings(self, data: dict[str, Any]) -> WorkflowSettings:
        """Parse workflow settings."""
        git_data = data.get("git", {})
        git = GitSettings(
            enabled=git_data.get("enabled", False),
            worktree=git_data.get("worktree", False),
            auto_commit=git_data.get("auto-commit", True),
            auto_pr=git_data.get("auto-pr", True),
            branch_prefix=git_data.get("branch-prefix", "agentic"),
        )
        return WorkflowSettings(
            max_retry=data.get("max-retry", 3),
            timeout_minutes=data.get("timeout-minutes", 60),
            track_progress=data.get("track-progress", True),
            autofix=data.get("autofix", "none"),
            terminal_output=data.get("terminal-output", "base"),
            bypass_permissions=data.get("bypass-permissions", False),
            git=git,
        )

    def _parse_variable(self, data: dict[str, Any]) -> Variable:
        """Parse a variable definition."""
        return Variable(
            name=self._required(data, "name"),
            type=data.get("type", "string"),
            required=data.get("required", True),
            default=data.get("default"),
            description=data.get("description", ""),
        )

    def _parse_step(self, data: dict[str, Any]) -> StepDefinition:
        """Parse a step definition."""
        name = self._required(data, "name")
        type_str = self._required(data, "type")

        try:
            step_type = StepType(type_str)
        except ValueError:
            valid = [t.value for t in StepType]
            raise WorkflowParseError(
                f"Invalid step type: {type_str}. Valid types: {valid}"
            ) from None

        step = StepDefinition(name=name, type=step_type)

        if step_type == StepType.PROMPT:
            step.prompt = data.get("prompt")
            step.agent = data.get("agent")
        elif step_type == StepType.COMMAND:
            step.command = data.get("command")
            step.args = data.get("args", {})
        elif step_type == StepType.PARALLEL:
            step.steps = [self._parse_step(s) for s in data.get("steps", [])]
            step.merge_strategy = data.get("merge-strategy", "wait-all")
            step.merge_mode = data.get("merge-mode", "independent")
        elif step_type == StepType.CONDITIONAL:
            step.condition = data.get("condition")
            step.then_steps = [self._parse_step(s) for s in data.get("then", [])]
            step.else_steps = [self._parse_step(s) for s in data.get("else", [])]
        elif step_type == StepType.RECURRING:
            step.max_iterations = data.get("max-iterations", 3)
            step.until = data.get("until")
            step.steps = [self._parse_step(s) for s in data.get("steps", [])]
        elif step_type == StepType.WAIT_FOR_HUMAN:
            step.message = data.get("message")
            step.polling_interval = data.get("polling-interval", 15)
            step.on_timeout = data.get("on-timeout", "abort")
            step.step_timeout_minutes = data.get("timeout-minutes", 5)

        step.model = data.get("model", "sonnet")
        step.step_timeout_minutes = data.get("timeout-minutes", step.step_timeout_minutes)
        step.step_max_retry = data.get("max-retry")
        step.on_error = data.get("on-error", "retry")
        step.checkpoint = data.get("checkpoint", False)
        step.depends_on = data.get("depends-on")

        return step

    def _parse_output(self, data: dict[str, Any]) -> OutputDefinition:
        """Parse an output definition."""
        return OutputDefinition(
            name=self._required(data, "name"),
            template=self._required(data, "template"),
            path=self._required(data, "path"),
            when=data.get("when", "completed"),
        )

File: src/parser.ts
/** YAML workflow parser. */

import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import yaml from "js-yaml";
import type {
	OutputDefinition,
	StepDefinition,
	StepType,
	Variable,
	WorkflowDefinition,
	WorkflowSettings,
} from "./types.js";
import { VALID_STEP_TYPES } from "./types.js";

export class WorkflowParseError extends Error {
	constructor(message: string) {
		super(message);
		this.name = "WorkflowParseError";
	}
}

export class WorkflowParser {
	basePath: string;

	constructor(basePath?: string) {
		this.basePath = basePath ?? process.cwd();
	}

	parseFile(filePath: string): WorkflowDefinition {
		if (!existsSync(filePath)) {
			throw new WorkflowParseError(`Workflow file not found: ${filePath}`);
		}

		let data: unknown;
		try {
			const content = readFileSync(filePath, "utf-8");
			data = yaml.load(content);
		} catch (e: unknown) {
			if (e instanceof WorkflowParseError) throw e;
			const msg = e instanceof Error ? e.message : String(e);
			throw new WorkflowParseError(`Invalid YAML: ${msg}`);
		}

		this.basePath = path.dirname(filePath);
		return this.parseDict(data);
	}

	parseString(content: string): WorkflowDefinition {
		let data: unknown;
		try {
			data = yaml.load(content);
		} catch (e: unknown) {
			const msg = e instanceof Error ? e.message : String(e);
			throw new WorkflowParseError(`Invalid YAML: ${msg}`);
		}
		return this.parseDict(data);
	}

	parseDict(data: unknown): WorkflowDefinition {
		if (!data || typeof data !== "object" || Array.isArray(data)) {
			throw new WorkflowParseError("Workflow must be a dictionary");
		}

		const dict = data as Record<string, unknown>;
		const name = this.required(dict, "name") as string;
		const version = (dict.version as string) ?? "1.0";
		const description = (dict.description as string) ?? "";

		const settings = this.parseSettings((dict.settings as Record<string, unknown>) ?? {});
		const variables = ((dict.variables as unknown[]) ?? []).map((v) =>
			this.parseVariable(v as Record<string, unknown>),
		);
		const steps = ((dict.steps as unknown[]) ?? []).map((s) =>
			this.parseStep(s as Record<string, unknown>),
		);
		const outputs = ((dict.outputs as unknown[]) ?? []).map((o) =>
			this.parseOutput(o as Record<string, unknown>),
		);

		return { name, version, description, settings, variables, steps, outputs };
	}

	private required(data: Record<string, unknown>, key: string): unknown {
		if (!(key in data)) {
			throw new WorkflowParseError(`Missing required field: ${key}`);
		}
		return data[key];
	}

	private parseSettings(data: Record<string, unknown>): WorkflowSettings {
		return {
			maxRetry: (data["max-retry"] as number) ?? 3,
			timeoutMinutes: (data["timeout-minutes"] as number) ?? 60,
			trackProgress: (data["track-progress"] as boolean) ?? true,
			autofix: (data.autofix as string) ?? "none",
			terminalOutput: (data["terminal-output"] as string) ?? "base",
			bypassPermissions: (data["bypass-permissions"] as boolean) ?? false,
			strictMode: (data["strict-mode"] as boolean) ?? false,
			model: (data.model as string) ?? null,
			runtime: (data.runtime as string) ?? null,
			requiredTools: (data["required-tools"] as string[]) ?? [],
		};
	}

	private parseVariable(data: Record<string, unknown>): Variable {
		return {
			name: this.required(data, "name") as string,
			type: (data.type as string) ?? "string",
			required: (data.required as boolean) ?? true,
			default: data.default,
			description: (data.description as string) ?? "",
		};
	}

	private parseStep(data: Record<string, unknown>, parentIsParallel = false): StepDefinition {
		const name = this.required(data, "name") as string;
		const typeStr = this.required(data, "type") as string;

		if (!VALID_STEP_TYPES.includes(typeStr as StepType)) {
			throw new WorkflowParseError(
				`Invalid step type: ${typeStr}. Valid types: ${JSON.stringify(VALID_STEP_TYPES)}`,
			);
		}

		const stepType = typeStr as StepType;

		if (parentIsParallel && stepType === "parallel") {
			throw new WorkflowParseError(
				`Nested parallel steps are not allowed: '${name}' is a parallel step inside another parallel step`,
			);
		}

		const step: StepDefinition = {
			name,
			type: stepType,
			prompt: null,
			agent: null,
			steps: [],
			condition: null,
			thenSteps: [],
			elseSteps: [],
			maxIterations: 5,
			completionPromise: null,
			message: null,
			pollingInterval: 15,
			onTimeout: "abort",
			model: null,
			runtime: null,
			stepTimeoutMinutes: null,
			stepMaxRetry: null,
			onError: "retry",
			checkpoint: false,
			dependsOn: null,
		};

		if (stepType === "prompt") {
			step.prompt = (data.prompt as string) ?? null;
			step.agent = (data.agent as string) ?? null;
		} else if (stepType === "parallel") {
			step.steps = ((data.steps as unknown[]) ?? []).map((s) =>
				this.parseStep(s as Record<string, unknown>, true),
			);
		} else if (stepType === "serial") {
			step.steps = ((data.steps as unknown[]) ?? []).map((s) =>
				this.parseStep(s as Record<string, unknown>),
			);
		} else if (stepType === "conditional") {
			step.condition = (data.condition as string) ?? null;
			step.thenSteps = ((data.then as unknown[]) ?? []).map((s) =>
				this.parseStep(s as Record<string, unknown>),
			);
			step.elseSteps = ((data.else as unknown[]) ?? []).map((s) =>
				this.parseStep(s as Record<string, unknown>),
			);
		} else if (stepType === "ralph-loop") {
			step.prompt = (data.prompt as string) ?? null;
			step.maxIterations = (data["max-iterations"] as number) ?? 5;
			step.completionPromise = (data["completion-promise"] as string) ?? null;
		} else if (stepType === "wait-for-human") {
			step.message = (data.message as string) ?? null;
			step.pollingInterval = (data["polling-interval"] as number) ?? 15;
			step.onTimeout = (data["on-timeout"] as string) ?? "abort";
			step.stepTimeoutMinutes = (data["timeout-minutes"] as number) ?? 5;
		}

		step.model = (data.model as string) ?? null;
		step.runtime = (data.runtime as string) ?? null;
		step.stepTimeoutMinutes = (data["timeout-minutes"] as number) ?? step.stepTimeoutMinutes;
		step.stepMaxRetry = (data["max-retry"] as number) ?? null;
		step.onError = (data["on-error"] as string) ?? "retry";
		step.checkpoint = (data.checkpoint as boolean) ?? false;
		step.dependsOn = (data["depends-on"] as string) ?? null;

		return step;
	}

	private parseOutput(data: Record<string, unknown>): OutputDefinition {
		return {
			name: this.required(data, "name") as string,
			template: this.required(data, "template") as string,
			path: this.required(data, "path") as string,
			when: (data.when as string) ?? "completed",
		};
	}
}
